torch/_refs/__init__.py

```diff
@@ -1655,48 +1655,108 @@ def item(a: TensorLikeType) -> NumberType:
     return number_type(prims.item(a))
 
 
-# TODO: extend to device and layout similar to
-# aten/src/ATen/native/TensorConversions.cpp:to_will_alias
 def _to_will_alias(
     a: TensorLikeType,
-    dtype: torch.dtype,
-    copy: bool,
-    memory_format: torch.memory_format,
+    device: Optional[torch.device],
+    dtype: Optional[torch.dtype],
+    copy: Optional[bool],
+    memory_format: Optional[torch.memory_format],
+    layout: Optional[torch.layout],
 ) -> bool:
     return (
         not copy
-        and a.dtype == dtype
+        and (device is None or a.device == device)
+        and (dtype is None or a.dtype == dtype)
+        and (layout is None or a.layout == layout)
         and (
-            memory_format == torch.preserve_format
-            or a.is_contiguous(memory_format=memory_format)
+            memory_format is None
+            or memory_format == torch.preserve_format
+            or utils.is_contiguous_for_memory_format(a, memory_format=memory_format)
         )
     )
 
 
-# TODO: there are multiple overloads of to, but we only support one here
-# aten/src/ATen/native/native_functions.yaml lists all overloads
-# Missing overloads:
-# to.dtype_layout
-# to.device
-# to.other
-# https://github.com/pytorch/pytorch/issues/84264
+def _to_dispatch(
+    device=None,
+    dtype=None,
+    non_blocking=None,
+    copy=None,
+    memory_format=None,
+    layout=None,
+    pin_memory=None,
+):
+    if isinstance(device, TensorLike):
+        # overload to `to.other`
+        if isinstance(copy, torch.memory_format):
+            memory_format = copy
+        # we need to make sure that dtype is bool, otherwise, non_blocking could be passed as kwarg
+        if isinstance(non_blocking, bool) and isinstance(dtype, bool):
+            copy = non_blocking
+        if isinstance(dtype, bool):
+            non_blocking = dtype
+        # extract tensor options
+        other = device
+        device = other.device
+        dtype = other.dtype
+        layout = other.layout
+        pin_memory = other.pin_memory
+    elif isinstance(device, torch.dtype):
+        # overload to `to.dtype`
+        if isinstance(copy, torch.memory_format):
+            memory_format = copy
+        # we need to make sure that dtype is bool, otherwise, non_blocking could be passed as kwarg
+        if isinstance(non_blocking, bool) and isinstance(dtype, bool):
+            copy = non_blocking
+        if isinstance(dtype, bool):
+            non_blocking = dtype
+        dtype = device
+        layout = None
+        pin_memory = None
+        device = None
+    # overload to `to.device`
+    # do nothing because signature of 'to' matches
+    # overload to `to.dtype_layout`
+    # do nothing because the `to.dtype_layout` takes all kwargs.
+    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory
+
+
 def to(
```

Comment:
It's unfortunate that `to.device` overload takes 6 positional arguments and `to.dtype`, `to.other` take 5 positional arguments.

Reply:
The `sample_inputs_to` function I added is not testing that.
```
diff --git a/torch/testing/_internal/common_methods_invocations.py b/torch/testing/_internal/common_methods_invocations.py
index 9a9d8c7a62..8b2e0b2834 100644
--- a/torch/testing/_internal/common_methods_invocations.py
+++ b/torch/testing/_internal/common_methods_invocations.py
@@ -3898,31 +3898,16 @@ def sample_inputs_to(op_info, device, dtype, requires_grad, **kwargs):

     # to.device overload
     for device, nb, cp, mem_f in product(devices, [True, False], [True, False], memory_formats):
-        kwargs = {
-            "non_blocking": nb,
-            "copy": cp,
-            "memory_format": mem_f,
-        }
-        yield SampleInput(make_arg((S, S, S, S)), args=(device, torch.float64,), kwargs=kwargs)
+        yield SampleInput(make_arg((S, S, S, S)), args=(device, torch.float64, nb, cp, mem_f))

     # to.dtype overload
     for nb, cp, mem_f in product([True, False], [True, False], memory_formats):
-        kwargs = {
-            "non_blocking": nb,
-            "copy": cp,
-            "memory_format": mem_f,
-        }
-        yield SampleInput(make_arg((S, S, S, S)), args=(torch.float64,), kwargs=kwargs)
+        yield SampleInput(make_arg((S, S, S, S)), args=(torch.float64, nb, cp, mem_f))

     # to.other overload
     for device, nb, cp, mem_f in product(devices, [True, False], [True, False], memory_formats):
-        kwargs = {
-            "non_blocking": nb,
-            "copy": cp,
-            "memory_format": mem_f,
-        }
         other = make_arg((S, S, S, S), dtype=torch.float64, device=device)
-        yield SampleInput(make_arg((S, S, S, S)), args=(other,), kwargs=kwargs)
+        yield SampleInput(make_arg((S, S, S, S)), args=(other, nb, cp, mem_f))
```

Reply:
To be fair, I don't think we have comprehensive tests after the changes neither. There could also be combinations of positional + kw args, which affects how dispatching in python is handling this. hahaha.

But I agree that, knowing that the dispatch is more bug prone on mapping positional arguments, changes here to the tests looks like a good idea. Will update.

Reply:
O.o Looks like the `to.device` is not exposed in python. Which seems strange as native_functions.yaml seems to have that entry. But I also don't know how aten handles overload/dispatch in python, I guess if frontend doesn't have that, it's safe to just not test it :rofl: 

```
In [3]: x.to("cuda:0", torch.float16, False, False, None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 x.to("cuda:0", torch.float16, False, False, None)

TypeError: to() takes from 0 to 4 positional arguments but 5 were given
```